drivers/samsung/displays/md_series.cr

```diff
@@ -0,0 +1,500 @@
+        self[:volume]     = values[1]
+        self[:audio_mute] = values[2] == 1
+        self[:input]      = INPUTS.new(values[3]).to_s
+        check_power_state
+      when COMMAND::Panel_mute
+        self[:power] = value == 0
+        check_power_state
+      when COMMAND::Volume
+        self[:volume] = value
+        self[:audio_mute] = false if value > 0
+      when COMMAND::Brightness
+        self[:brightness] = value
+      when COMMAND::Input
+        self[:input] = INPUTS.new(value).to_s
+        # The input feedback behaviour seems to go a little odd when
+        # screen split is active. Ignore any input forcing when on.
+        unless self[:screen_split]
+          self[:input_stable] = self[:input] == self[:input_target]
+          switch_to(self[:input_target].as_s) unless self[:input_stable]
+        end
+      when COMMAND::Speaker
+        self[:speaker] = SPEAKERMODES.new(value).to_s
+      when COMMAND::Hard_off
+        self[:hard_off] = value == 0
+        self[:power] = false if self[:hard_off]
+      when COMMAND::Screen_split
+        self[:screen_split] = value >= 0
+      when COMMAND::Software_version
+        self[:software_version] = values.join
+      when COMMAND::Serial_number
+        self[:serial_number] = values.join
+      else
+        logger.debug { "Samsung responded with ACK: #{value}" }
+      end
+
+      task.try &.success
+    when RESPONSESTATUS::Nak
+      task.try &.abort("Samsung responded with NAK: #{hex}")
+    else
+      task.try &.abort("Samsung aborted with: #{hex}")
+    end
+  end
+
+  def check_power_state
+    return if self[:power_stable]?
+    if self[:power]? == self[:power_target]?
+      self[:power_stable] = true
+    else
+      power(self[:power_target].as_bool)
+    end
+  end
+
+  enum COMMAND
+    Status           = 0x00
+    Hard_off         = 0x11 # Completely powers off
+    Panel_mute       = 0xF9 # Screen blanking / visual mute
+    Volume           = 0x12
+    Contrast         = 0x24
+    Brightness       = 0x25
+    Sharpness        = 0x26
+    Colour           = 0x27
+    Tint             = 0x28
+    Red_gain         = 0x29
+    Green_gain       = 0x2A
+    Blue_gain        = 0x2B
+    Input            = 0x14
+    Mode             = 0x18
+    Size             = 0x19
+    Pip              = 0x3C # picture in picture
+    Auto_adjust      = 0x3D
+    Wall_mode        = 0x5C # Video wall mode
+    Safety           = 0x5D
+    Wall_on          = 0x84 # Video wall enabled
+    Wall_user        = 0x89 # Video wall user control
+    Speaker          = 0x68
+    Net_standby      = 0xB5 # Keep NIC active in standby
+    Eco_solution     = 0xE6 # Eco options (auto power off)
+    Auto_power       = 0x33
+    Screen_split     = 0xB2 # Tri / quad split (larger panels only)
+    Software_version = 0x0E
+    Serial_number    = 0x0B
+    Time             = 0xA7
+  end
+
+  private def do_send(command : String | Int, data : Int | Array(Int) = [] of UInt8, **options)
```

Comment:
Use the types here to help prevent error. `command` should only accept an `Command` enum.

Reply:
Hey @KimBurgess , I've made this change but there was a method in the ruby driver which allowed sending of custom commands via backoffice. Let me know if I should remove this though if it's not needed.

https://github.com/PlaceOS/drivers/blob/78c91c24b876507d49999657f36b0580ffa79214/drivers/samsung/displays/mdc_protocol.cr#L149-L154

Reply:
Yep, nuke it. Better to have fully defined commands when / where they're needed.

Reply:
No worries, nuked!
https://github.com/PlaceOS/drivers/pull/19/commits/52f408d53c0eedfd0201fd78dcdf27a7ec7ff8a6